src/QueryReflection/QueryReflection.php

```diff
@@ -99,6 +101,27 @@ public function getResultType(string $queryString, int $fetchType): ?Type
         return self::reflector()->getResultType($queryString, $fetchType);
     }
 
+    /**
+     * Determine if a query will be resolvable.
+     *
+     * - If yes, the query is a literal string.
+     * - If no, the query is a non-literal string or mixed type.
+     * - If maybe, the query is neither of the two.
+     *
+     * We will typically skip processing of queries that return no, which are
+     * likely part of a software abstraction layer that we know nothing about.
+     */
+    public function isResolvable(Expr $queryExpr, Scope $scope): TrinaryLogic
+    {
+        $type = $scope->getType($queryExpr);
+        if ($type->isLiteralString()->yes()) {
+            return TrinaryLogic::createYes();
+        }
+        $isStringOrMixed = $type->isSuperTypeOf(new StringType());
+
+        return $isStringOrMixed->negate();
```

Comment:
I would prefer something like

```suggestion
        if ($type instanceof MixedType) {
          return TrinaryLogic::createNo();
        }
        
        return $type->isString()->yes();
```

Reply:
I am pretty sure the logic in my suggestion is wrong, but I think you get the point.

otherwise the PR looks good, thank you

Reply:
So I'm actually having a bit of trouble with this logic. Do you think you could give me some pointers?

Basically, the logic that I'm trying to encapsulate is that if _any_ part of the string is a literal string, then we don't want to skip the query, because there is literal SQL that we should be checking. Then either it will resolve and be checked, or we'll warn that it doesn't resolve (with this PR). For example, I'd want the warning for the following:

```php
/** @var string $string */
$query = 'SELECT * FROM table WHERE ' . $string;
```
Pre-existing tests explicitly state that entirely non-literal string and mixed types should _not_ be checked, e.g.:

```php
/** @var string $string */
$query = $string;
```
with the reasoning that this is likely part of a s/w abstraction layer that phpstan-dba won't know about, which is reasonable.

But how does one distinguish between `$string` and `$literal . $string`? I was thinking that the former is a StringType and the latter is an IntersectionType, but many non-literal string variables are an IntersectionType with an Accessory, e.g. if `$string` is a `non-empty-string`. Is there any way to robustly do what I'm trying to do with the current phpstan type system, or is my idea fundamentally flawed?

To be clear, the edge cases that my attempts fail at look like, e.g.
```php
/** @var non-empty-string $string */
$query = $string;
```
I still think this is an important feature, especially for people modernizing old codebases that didn't use prepared queries. Enabling this warning has uncovered many unresolvable queries that I didn't find by hand (including ones that were already migrated to prepared statements, but unexpectedly used a non-literal string in the query construction).

If there is no such solution, I would honestly be totally happy with an option to just warn for _all_ unresolvable queries (with the idea that it's safer to explicitly ignore an unwanted warning than silently miss a bunch of real errors). But I look to you for guidance here, before I go too far off the deep end. :)

Thanks again for your time!

Reply:
I agree with the goal and the results of this PR.

my suggestion was more about how to make the logic in `isResolvable` more redable without changing what it is doing.

but since you added extensive coverage, I will just merge it, after merge conflicts are resolved.

sorry for the noise